### PR TITLE
Qualification tool output recommendations on a per sql query basis

### PR DESCRIPTION
**Pull request by tgravescs**
fixes https://github.com/NVIDIA/spark-rapids/issues/5502

This adds in an option to allow a user to get the recommendations at the per sql query level in addition to the application level.  If an application has a lot of sql queries, this lets you narrow it down further to figure out which queries might benefit from being on the GPU the most.

2 new output files rapids_4_spark_qualification_output_persql.log and rapids_4_spark_qualification_output_persql.csv for text format and then csv format, contents are the same between them for now.

output looks like:
```
==========================================================================================================================================================================================================================================================
|                              App Name|             App ID|SQL ID|                                            SQL Description|SQL DF Duration|GPU Opportunity|Estimated GPU Duration|Estimated GPU Speedup|Estimated GPU Time Saved|      Recommendation|
==========================================================================================================================================================================================================================================================
|Rapids Spark Profiling Tool Unit Tests|local-1622043423018|     1|                   count at QualificationInfoUtils.scala:94|           7143|           6719|               2716.79|                 2.62|                  4426.2|Strongly Recommended|
|Rapids Spark Profiling Tool Unit Tests|local-1622043423018|     3|                   count at QualificationInfoUtils.scala:94|           2052|           1660|                958.28|                 2.14|                 1093.71|         Recommended|
|Rapids Spark Profiling Tool Unit Tests|local-1622043423018|     2|                   count at QualificationInfoUtils.scala:94|           1933|           1551|                 911.3|                 2.12|                 1021.69|         Recommended|
|                           Spark shell|local-1651187225439|     0|                                       show at <console>:26|            498|            249|                 373.5|                 1.33|                   124.5|         Recommended|

```

2 new options added, one to turn this output on and one to control the size of the sql description printed. Default is 100 there.  The normal options qualification tool take apply to this table as well. Such as order and limiting number of rows.

This does not contain the UI changes.

### CODE REVIEW

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
looks like I didn't upmerge to latest, will update shortly

**Comment by tgravescs** [comment]
build

**Review by nartal1 — COMMENTED** [review]
Overall looks good. Just a question on `perSqlInfo`

File: tools/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala [hunk 1]
```diff
@@ -361,12 +363,30 @@ class QualificationAppInfo(
       val allSQLDurations = sqlIdToInfo.map { case (_, info) =>
         info.duration.getOrElse(0L)
       }
+
+      val appName = appInfo.map(_.appName).getOrElse("")
+      val perSqlInfos = if (reportSqlLevel) {
+        Some(origPlanInfos.flatMap { pInfo =>
```
[the remark this answers is not in the document]
Comment (tgravescs) [reply tools/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala:369]: yes good catch

**Comment by tgravescs** [comment]
build

**Review by nartal1 — APPROVED** [review]

**Review by amahussein — CHANGES_REQUESTED** [review]
The CSV file format may get corrupted by the SQL description.
The SQL descriptions seem to include ","
```
Spark shell,app-20210509200722-0001,105,Execution: ss_max-v2.4, iteration: 1, StandardRun=true,335,0,335.0,1.0,0.0,Not Recommended
```

The above row has two extra columns compared to other ones.

**Comment by tgravescs** [comment]
ah yep, needs to replace those

**Comment by tgravescs** [comment]
build

**Comment by pxLi** [comment]
unrelated `Unable to find py4j in` failed the CI intermittently. let me retrigger

**Comment by pxLi** [comment]
build

**Review by amahussein — APPROVED** [review]
This fixes the bug. It is concerning that the fix did not require any change in the unit tests.
Later, we may consider covering this in the unit tests.

LGTME.

**Comment by tgravescs** [comment]
I'm going to merge this and followup with a test in separate pr.

**Pull request merged**